packages/core/workflows-sdk/src/utils/composer/helpers/create-step-handler.ts

```diff
@@ -0,0 +1,83 @@
+import {CreateWorkflowComposerContext, StepExecutionContext} from "../type"
+import {WorkflowStepHandlerArguments} from "@medusajs/orchestration"
+import {resolveValue} from "./resolve-value"
+import {StepResponse} from "./step-response"
+import {deepCopy, OrchestrationUtils} from "@medusajs/utils"
+
+export function createStepHandler(
+  this: CreateWorkflowComposerContext,
+  { stepName, input, invokeFn, compensateFn }
+) {
+  const handler = {
+    invoke: async (stepArguments: WorkflowStepHandlerArguments) => {
+      const metadata = stepArguments.metadata
+      const idempotencyKey = metadata.idempotency_key
+
+      stepArguments.context!.idempotencyKey = idempotencyKey
+      const executionContext: StepExecutionContext = {
+        workflowId: metadata.model_id,
+        stepName: metadata.action,
+        action: "invoke",
+        idempotencyKey,
+        attempt: metadata.attempt,
+        container: stepArguments.container,
+        metadata,
+        eventGroupId:
+          stepArguments.transaction.getFlow()?.metadata?.eventGroupId ??
+          stepArguments.context!.eventGroupId,
+        transactionId: stepArguments.context!.transactionId,
+        context: stepArguments.context!,
+      }
```

Comment:
suggestion: wdyt if you extract that to a separate function (e.g stepContextBuilder) which takes the action and stepArguments and build the step context, then we can use it here and in the compensate as well and be sure that only one place is building this context

Reply:
I will leave the existing logic cleanup to a separate PR :)